### ISSUE TITLE
feat: DynamicTruncate op for runtime-determined tensor slicing

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#691]
## Summary

Add a `DynamicTruncate` op that slices a tensor along an axis to a runtime-determined size (given as a scalar tensor input). This enables adaptive truncated SVD inside a single traced graph without eval barriers.

## Motivation

In TN algorithms, truncated SVD determines the output rank from singular values at runtime:
```
rank = count(S > rtol * S[0])
rank = min(rank, max_rank)
U_r = U[:, :rank], S_r = S[:rank], V_r = V[:rank, :]
```

In a traced/graph system, this is problematic:
- `dynamic_slice` requires static `slice_sizes` (compile-time constant)
- Getting rank requires eval barrier (S values needed) → per-step eval → O(N²) cost or forced O(L) checkpoints
- Mask approach (zero out small S, keep full shape) wastes compute: O(k_full/actual_rank) overhead

`DynamicTruncate` solves this by making the slice size a graph input (runtime scalar), enabling the entire truncation to be expressed as graph ops without eval barriers.

### Comparison with JAX/PyTorch

- **PyTorch eager**: adaptive truncation works naturally (no graph constraints). With `torch.utils.checkpoint`, √L memory is achievable.
- **JAX JIT**: static shapes required → mask to k_full or fixed rank. Cannot combine adaptive truncation with optimal checkpoint spacing.
- **tenferro-rs + DynamicTruncate**: brings eager-level flexibility to graph system. Enables adaptive truncation + sparse √L checkpoints + compile cache + graph-level fusion.

### Op definition

```rust
// In StdTensorOp
DynamicTruncate { axis: usize }
// input[0]: tensor to truncate
// input[1]: scalar tensor — runtime size (number of elements to keep along axis)
// output: tensor sliced to first input[1] elements along axis
// output shape: determined at runtime by input[1]'s value
```

### Graph construction for truncated SVD

The rank determination is expressed entirely as existing + new graph ops:

```
input matrix A
    ↓
SVD(A) → U(m, k), S(k), Vt(k, n)                     [existing: Svd op]
    ↓
threshold = reduce_max(S, [0]) * rtol_constant         [existing: ReduceMax + Mul + Constant]
    ↓
mask = compare_gt(S, broadcast(threshold))              [existing: Compare + BroadcastInDim]
    ↓
rank = reduce_sum(mask, [0])                            [existing: ReduceSum — gives float count]
    ↓
rank = minimum(rank, max_rank_constant)                 [existing: Minimum + Constant]
    ↓
U_r = DynamicTruncate(U, rank, axis=1)                 [NEW]
S_r = DynamicTruncate(S, rank, axis=0)                 [NEW]
Vt_r = DynamicTruncate(Vt, rank, axis=0)               [NEW]
```

All ops except `DynamicTruncate` already exist. The rank flows through the graph as a scalar tensor value. No eval barrier needed.

### Execution

```rust
// In eval_exec_ir (exec.rs), as a new ExecOp variant:
ExecOp::DynamicTruncate { axis } => {
    let input = get(&slots, &inst.input_slots, 0)?;
    let size_tensor = get(&slots, &inst.input_slots, 1)?;
    let size = extract_scalar_usize(size_tensor)?;
    // Slice input along axis from 0 to size
    let result = slice_along_axis(input, axis, 0..size)?;
    result
}
```

Output shape is runtime-determined. The exec engine handles this (slots store `Option<Tensor>`, no pre-allocated shape). Downstream ops (NaryEinsum, DotGeneral) resolve shapes at exec time.

### StableHLO lowering

```rust
// compiler.rs
StdTensorOp::DynamicTruncate { axis } => StableHloOp::CustomCall {
    target: format!("dynamic_truncate_{axis}"),
}

// For future XLA backend: lower to pad-to-bound + SetDimensionSize
// or graph break + static slice (same as PyTorch torch.compile guard_bool)
```

### AD rules

**linearize (JVP):**
```rust
// tangent of DynamicTruncate = DynamicTruncate on the tangent
// (same truncation applied to tangent)
fn linearize_dynamic_truncate(builder, primal_in, tangent_in, axis) {
    let Some(dt) = tangent_in[0] else { return vec![None] };
    // size input (primal_in[1]) has zero tangent (discrete, non-differentiable)
    let out = builder.add_op(
        StdTensorOp::DynamicTruncate { axis },
        vec![ValRef::Local(dt), ValRef::External(primal_in[1].clone())],
        OpMode::Linear { active_mask: vec![true, false] },
    );
    vec![Some(out[0])]
}
```

**transpose (VJP):**
```rust
// Adjoint of truncate = pad with zeros to original size
// Original size comes from the primal input's shape along axis
fn transpose_dynamic_truncate(builder, cotangent_out, inputs, mode, axis) {
    let Some(ct) = cotangent_out[0] else { return vec![None, None] };
    // Pad cotangent back to original input size along axis
    // Use DynamicPad or Pad op with appropriate zero-padding config
    // The original size can be obtained from the primal input reference
    let padded = pad_along_axis(builder, ct, inputs[0], axis);
    vec![Some(padded), None]  // no gradient for size input
}
```

**Note:** The VJP requires knowing the original (pre-truncation) size to pad back. This is available from the primal input tensor's shape. A new `DynamicPad` op may be needed, or we can use the existing `Pad` op with a config derived from the primal input shape at exec time.

### Prerequisites

- [ ] `Compare` op exposed as TracedTensor method (currently private in linalg_api.rs)
- [ ] `Minimum`/`Maximum` AD rules (currently `todo!()` — panic if AD attempted). For DynamicTruncate, these are in the rank determination path which is stop-gradient, so AD rules are NOT required for the initial implementation. However, they should be added for general use.
- [ ] `ReduceMax` for extracting S[0] (or use `Slice` to extract first element)

### Interaction with checkpoint (#690)

DynamicTruncate removes per-step eval barriers, enabling flexible checkpoint placement:

| Approach | Checkpoints | Memory |
|---|---|---|
| Checkpoint at every SVD (no DynamicTruncate) | L (forced) | O(L) |
| DynamicTruncate + √L checkpoints | √L (optimal) | O(√L) |

For L=10000, tensor_size=8MB: O(L)=80GB vs O(√L)=800MB.

### Testing

- [ ] DynamicTruncate forward: truncate [1,2,3,4,5] with size=3 → [1,2,3]
- [ ] DynamicTruncate with rank from compare+reduce_sum: full truncated SVD pipeline
- [ ] AD through DynamicTruncate: JVP (finite-diff verified)
- [ ] AD through DynamicTruncate: VJP (finite-diff verified)
- [ ] AD through full truncated SVD pipeline (SVD + rank determination + DynamicTruncate)
- [ ] Downstream NaryEinsum with DynamicTruncate outputs (runtime shapes)
- [ ] L-step sweep without eval barriers: verify single graph, single eval
- [ ] Combine with checkpoint (#690): √L checkpoints + DynamicTruncate

## Related

- #690 — checkpoint mechanism for TracedTensor
- #681 — NaryEinsum op (enables downstream ops with runtime shapes)
- #682 — symbolic shape_hint (Vec<SymDim>)
- `docs/design/symbolic-shape-and-tenferro-migration.md` — Phase 4 design

**Comment by shinaoka** [comment]
Implemented by #692 (merged). `DynamicTruncate` op with `PadToMatch` adjoint, full AD rules (linearize + transpose), HVP verified.
